app/js/authentication_controller.js

```diff
@@ -0,0 +1,138 @@
+'use strict';
+
+/* global User, Utils */
+
+(function(exports) {
+  function update_error_list(list, errors) {
+    list.innerHTML = '';
+    if (errors) {
+      for (var i in errors) {
+        var error = document.createElement('li');
+        error.appendChild(document.createTextNode(errors[i]));
+        list.appendChild(error);
+      }
+    }
+  }
+
+  function authenticate(evt) {
+    evt.preventDefault();
+
+    var form = document.getElementById('login');
+    var username = form.elements.username.value;
+    var password = form.elements.password.value;
+
+    var promise = User.authenticate_user(username, password);
+
+    Utils.toggle_spinner();
+
+    promise.then(function(token) {
+      User.set_user(username, password, token).then(function() {
+        Utils.toggle_spinner();
+      });
+    }).catch(function(response) {
+      var errors = JSON.parse(response);
+
+      update_error_list(document.getElementById('login_errors'),
+        errors.non_field_errors);
+
+      update_error_list(document.getElementById('username_errors'),
+        errors.username);
+
+      update_error_list(document.getElementById('password_errors'),
+        errors.password);
+
+      Utils.toggle_spinner();
+    });
+  }
+
+  function request_password_reset(evt) {
+    evt.preventDefault();
+
+    var form = document.getElementById('pw_reset');
+    var username = form.elements.username.value;
+
+    var promise = User.request_password_reset(username);
+
+    Utils.toggle_spinner();
+
+    promise.then(function() {
+      var success = document.getElementById('success_message');
+      form.classList.add('hide');
+      success.classList.remove('hide');
+      Utils.toggle_spinner();
+    }).catch(function(response) {
+      update_error_list(document.getElementById('pw_reset_errors'),
+        ['User not found.']);
+      Utils.toggle_spinner();
+    });
+  }
+
+  function register_user(evt) {
+    evt.preventDefault();
+
+    var form = document.getElementById('register');
+    var email = form.elements.email.value;
+    var username = form.elements.username.value;
+    var password = form.elements.password.value;
+
+    var promise = User.register(username, password, email);
+
+    Utils.toggle_spinner();
+
+    promise.then(function() {
+      User.authenticate_user(username, password).then(function(token) {
+        User.set_user(username, password, token).then(function() {
```

Comment:
Also, this might be useful when you move that to `User`, Promises allow you to not go crazy with the pyramid of doom. Hence a cleaner to write this is:
```js
User.register(username, password, email).then(function() {
  return User.authenticate_user(username, password);
}).then(function(token) {
  return User.set_user(username, password, token);
}, function(response) {
  // Update error UI
}).then(function() {
  // Update success UI
});